src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -27,139 +27,507 @@
+    absl::flat_hash_set<ResourceID> res;
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      if (predefined_resources_[i].size() > 0) {
+        res.insert(ResourceID(i));
+      }
+    }
+    for (auto &entry : custom_resources_) {
+      res.insert(ResourceID(entry.first));
+    }
+    return res;
+  }
+
+  /// Return the number of resources in this set.
+  size_t Size() const {
+    size_t size = custom_resources_.size();
+    for (size_t i = 0; i < PredefinedResourcesEnum_MAX; i++) {
+      if (predefined_resources_[i].size() > 0) {
+        size++;
+      }
+    }
+    return size;
+  }
+
+  /// Check whether this set is empty.
+  bool IsEmpty() const { return Size() == 0; }
+
+  bool operator==(const TaskResourceInstances &other) const {
+    for (size_t i = 0; i < PredefinedResourcesEnum_MAX; i++) {
+      if (!FixedPointVectorEqual(this->predefined_resources_[i],
+                                 other.predefined_resources_[i])) {
+        return false;
+      }
+    }
+    return this->custom_resources_ == other.custom_resources_;
+  }
+
+  /// Return a ResourceRequest with the aggregated per-instance values.
+  ResourceRequest ToResourceRequest() const {
+    ResourceRequest resource_request;
+    for (auto resource_id : ResourceIds()) {
+      resource_request.Set(resource_id, Sum(resource_id));
     }
-  };
-  /// Clear only the CPU instances field.
-  void ClearCPUInstances();
-  /// Check whether there are no resource instances.
-  bool IsEmpty() const;
+    return resource_request;
+  }
+
   /// Returns human-readable string for these resources.
-  [[nodiscard]] std::string DebugString() const;
+  [[nodiscard]] std::string DebugString() const {
+    std::stringstream buffer;
+    buffer << "{";
+    bool first = true;
+    for (auto resource_id : ResourceIds()) {
+      if (!first) {
+        buffer << ", ";
+      }
+      first = false;
+      buffer << resource_id.Binary() << ": "
+             << FixedPointVectorToString(Get(resource_id));
+    }
+    buffer << "}";
+    return buffer.str();
+  }
+
+  std::string SerializeAsJson() const {
+    bool has_added_resource = false;
+    std::stringstream buffer;
+    buffer << "{";
+    for (size_t i = 0; i < PredefinedResourcesEnum_MAX; i++) {
+      std::vector<FixedPoint> resource = predefined_resources_[i];
+      if (resource.empty()) {
+        continue;
+      }
+      if (has_added_resource) {
+        buffer << ",";
+      }
+      std::string resource_name = ResourceID(i).Binary();
+      buffer << "\"" << resource_name << "\":";
+      if (!ResourceID(i).IsUnitInstanceResource()) {
+        buffer << resource[0];
+      } else {
+        buffer << "[";
+        for (size_t i = 0; i < resource.size(); i++) {
+          buffer << resource[i];
+          if (i < resource.size() - 1) {
+            buffer << ", ";
+          }
+        }
+        buffer << "]";
+      }
+      has_added_resource = true;
+    }
+    // TODO (chenk008): add custom_resources_
```

Comment:
Here we can iterate all resource ids to remove this TODO?

Reply:
I'll leave it to @chenk008 
For this PR, I'd like to only do refactor and minimize behavior changes.